openedx/openedx-platform
# feat: add grading method support

## Description
This PR adds a new field in the Problem settings for choosing a **Grading Method**. Currently, the only Grading Method is the Last Score. From now on, the new grading methods allow are:

1. **Last Score _(Default)_**: The last score made is taken for grading.
2. **First Score**: The first score made is taken for grading.
3. **Highest Score**: The highest score made is taken for grading.
4. **Average Score**: The average of all scores made is taken for grading.

> [!IMPORTANT]  
> This feature is only available if you enable the following feature flag:
> ```bash
> ENABLE_GRADING_METHOD_IN_PROBLEMS
> ```

## Supporting information
These changes are part of the effort made to implement [Configurable grading method for problems with multiple attempts](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3920789505/Configurable+grading+method+for+problems+with+multiple+attempts)

## Dependencies
> [!NOTE]  
> This PR works with the Legacy Problem Editor interface, but if you use the new interface with [Course Authoring MFE](https://github.com/openedx/frontend-app-course-authoring/), you need the changes in [this PR](https://github.com/openedx/frontend-lib-content-components/pull/430).

### Demo

https://github.com/openedx/edx-platform/assets/64033729/74931434-c324-481d-b294-9a832d59b7eb

## Testing Instructions
1. Add in your environment this setting to active the feature: `FEATURES["ENABLE_GRADING_METHOD_IN_PROBLEMS"] = True`
2. Go to Studio and create a problem component from a Unit. e.g. `Advanced` > `Blank Problem`. In the settings you can add this example of a problem:

    ```xml
    <problem>
      <multiplechoiceresponse>
        <div>Grading Method Sample</div>
        <choicegroup>
          <choice correct="true">
            <div>Correct</div>
          </choice>
          <choice correct="false">
            <div>Incorrect</div>
          </choice>
          <choice correct="false">
            <div>Incorrect</div>
          </choice>
          <choice correct="false">
            <div>Incorrect</div>
          </choice>
        </choicegroup>
      </multiplechoiceresponse>
    </problem>
    ```
3. In component settings, you should see a new field: `Grading Method`. This field is a dropdown list with the Grading Methods mentioned above.
4. Choose a Grading Method and save changes.
5. From the LMS, answer the problem and check with the different Grading Methods. You can also check from the progress section.

### Rescoring
1. From Studio, edit the correct answer to the problem, or change the Grading Method.
2. From the LMS as an instructor, make the rescore of the problem in `STAFF DEBUG INFO` > ` Rescore Learner's Submission` 
3. The final score should have changed depending on the new answer or new Grading Method.
